### PR TITLE
REALMC-5872 merging topology changes to v1.4.1 driver

### PR DESCRIPTION
**Pull request by kpatel71716**

### CODE REVIEW

File: mongo/client.go [hunk 1]
```diff
@@ -776,7 +776,7 @@ func (c *Client) ListDatabaseNames(ctx context.Context, filter interface{}, opts
 //
 // Any error returned by the fn callback will be returned without any modifications.
 func WithSession(ctx context.Context, sess Session, fn func(SessionContext) error) error {
-	return fn(NewSessionContext(ctx, sess))
+	return fn(ContextWithSession(ctx, sess))
```
Comment (kpatel71716) [inline mongo/client.go]: I can leave the function named as NewSessionContext and rename the references in the baas repo instead (there are not that many). Not sure how we have previously done this but removing these renamings will keep us more 1 to 1 with the main driver

File: x/mongo/driver/topology/connection.go [hunk 2]
```diff
@@ -111,17 +111,17 @@ func (c *connection) connect(ctx context.Context) {
 	}
 	defer close(c.connectDone)
 
-	c.connectContextMutex.Lock()
+	c.mu.Lock()
```
Comment (kpatel71716) [inline x/mongo/driver/topology/connection.go]: I can revert these changes as well. I do not believe this variable is referenced outside of the driver code so there should be no impact

File: x/mongo/driver/topology/topology.go [hunk 3]
```diff
@@ -673,3 +673,41 @@ func (t *Topology) String() string {
 	}
 	return fmt.Sprintf("Type: %s, Servers: [%s]", desc.Kind, serversStr)
 }
+
+// isTopologyConsistent returns false when we have a replica set that claims to
+// have no primary but there exists a primary with all other nodes as secondaries. This
+// specifically works around HELP-13825.
+func (t *Topology) IsConsistent() bool {
```
Comment (edaniels) [inline x/mongo/driver/topology/topology.go]: we don't need this anymore since we don't use it

**Review by edaniels — APPROVED** [review]

File: x/mongo/driver/topology/topology.go (hunk 3 above)
Comment (kpatel71716) [reply x/mongo/driver/topology/topology.go]: Awesome! removed

**Pull request merged**